src/main/java/appeng/core/api/ApiPart.java

```diff
@@ -80,64 +90,79 @@ public void initFMPSupport()
+		return cache.getUnchecked( new CacheKey( baseClass, this.desc ) );
+	}
+
+	private Class<? extends AEBaseTile> generateCombinedClass( CacheKey cacheKey )
+	{
+		final Class<? extends AEBaseTile> parentClass;
+
+		// Get the list of interfaces that still need to be implemented beyond the current one
+		List<String> remainingInterfaces = cacheKey.getInterfaces().subList( 1, cacheKey.getInterfaces().size() );
 
-		if( this.tileImplementations.get( description ) != null )
+		// We are not at the root of the class hierarchy yet
+		if( !remainingInterfaces.isEmpty() )
+		{
+			CacheKey parentKey = new CacheKey( cacheKey.getBaseClass(), remainingInterfaces );
+			parentClass = cache.getUnchecked( parentKey );
+		}
+		else
 		{
-			return this.tileImplementations.get( description );
+			parentClass = cacheKey.getBaseClass();
 		}
 
-		String f = base;// TileCableBus.class.getName();
-		String Addendum = baseClass.getSimpleName();
-		Class<? extends AEBaseTile> myClass = baseClass;
+		// Which interface should be implemented in this layer?
+		String interfaceName = cacheKey.getInterfaces().get( 0 );
 
-		String path = f;
+		try
+		{
+			// This is the particular interface that this layer was registered for. Loading the class may fail if i.e. an API is broken or not present
+			// and in this case, the layer will be skipped!
+			Class<?> interfaceClass = Class.forName( interfaceName );
+			String layerImpl = this.interfaces2Layer.get( interfaceClass );
 
-		for( final String name : this.desc )
+			return this.getClassByDesc( parentClass, layerImpl );
+		}
+		catch( final Throwable t )
 		{
-			try
-			{
-				final String newPath = path + ';' + name;
-				myClass = this.getClassByDesc( baseClass.getSimpleName(), newPath, f, this.interfaces2Layer.get( Class.forName( name ) ) );
-				path = newPath;
-			}
-			catch( final Throwable t )
-			{
-				AELog.warn( "Error loading " + name );
-				AELog.debug( t );
-			}
-			f = myClass.getName();
+			AELog.warn( "Error loading " + interfaceName );
+			AELog.debug( t );
+			return parentClass;
 		}
 
-		this.tileImplementations.put( description, myClass );
-
-		return myClass;
 	}
 
-	private Class getClassByDesc( final String addendum, final String fullPath, final String root, final String next )
+	@SuppressWarnings( "unchecked" )
+	private Class<? extends AEBaseTile> getClassByDesc( Class<? extends AEBaseTile> baseClass, final String next )
 	{
-		if( this.roots.get( fullPath ) != null )
-		{
-			return this.roots.get( fullPath );
-		}
-
 		final ClassWriter cw = new ClassWriter( ClassWriter.COMPUTE_MAXS );
 		final ClassNode n = this.getReader( next );
 		final String originalName = n.name;
 
 		try
 		{
-			n.name = n.name + '_' + addendum;
-			n.superName = Class.forName( root ).getName().replace( ".", "/" );
+			n.name = n.name + '_' + baseClass.getSimpleName();
+			n.superName = baseClass.getName().replace( ".", "/" );
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Method appeng.core.api.ApiPart.getClassByDesc(Class, String) passes constant String of length 1 to character overridden method [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=fb-contrib%3AUCPM_USE_CHARACTER_PARAMETERIZED_METHOD)